### ISSUE TITLE
Word based suggestions stopped working properly being taken over by snippet suggestions

### ISSUE THREAD

**Issue by maciejmatczak** [microsoft/vscode#136500]
Issue Type: <b>Bug</b>

# Description

I am editing TCL files. TCL plugin (tcl|bit|0.4.0) wasn't updated recently at all. For this specific setup, I also have a `tcl.json` - set of ~10k snippets from 3rd party library. Everything worked great since months if not year[s].

Up to recently everything worked as expected:
 - I wrote some partial name of variable, hit CTRL-SPACE and I got either a word based completion or a snippet completion, depending which pattern match was better

Then, the only thing I am aware, I revisited my `tcl.json` snippet file, added one more. After VS Code reloaded snippets, I no longer get a proper word based suggestions, unless NONE of the snippets is matched (which is hard to do with some kind of fuzzy matching). It looks almost like VS Code was using some cached version (working properly), then rebuilt something.

- I type one letter - my completion consists only snippets
- I type further, being very close to the name of variable - still non of the word based suggestion appears, only list of snippets suggestions gets narrowed down

- if I type a combination that none of the snippet matches - then word suggestion somehow appears

Hitting ESC, CTRL-SPACE helps depending on a current input word and how it matches against snippets.

If I turn off snippet matching (`"editor.snippetSuggestions": "none"`), word based suggestion works fine. Setting it to "bottom" doesn't change anything (still doesn't work and I only get snippets).

It's like hitting some limiter that entirely turns off word based suggestions.

# Steps to reproduce

 - install `tcl|bit|0.4.0` https://marketplace.visualstudio.com/items?itemName=bitwisecook.tcl
 - CTRL-SHIFT-P, "Configuer User Snippets", pick language TCL, paste these snippets:
    ```json
    {
        "my something": {
            "prefix": "my-something",
            "body": "my::something ${1}\n${0}"
        },
        "my function": {
            "prefix": "my-function",
            "body": "my::function ${1} ${2}\n${0}"
        }
    }
    ```
 - open folder in VS Code, create `somefile.tcl`, paste in this content:
     ```tcl
    namespace eval firstProgram {
        proc hello { {myVar world} } {
            # try to print myVar (command in TCL is `puts`, value of variable is
            # reached with $)
            # after $ character press ctrl-space
    
            puts $
        }
    }
    ```

I can't workout the logic I see from editor's behavior.

1. If I input `puts $m<ctrl-space>`, I get auto suggestion triggered with snippets (sometimes triggers without ctrl-space); I should see words and snippets:
![image](https://user-images.githubusercontent.com/1786383/140505611-b1ebde36-ff27-4069-b18b-a2ee5f52c4c1.png)

2. If I input `puts $<ctrl-space>`, I see list of words:
![image](https://user-images.githubusercontent.com/1786383/140505730-f118bdb6-fec7-48e9-89dd-328ac12fab17.png)
  a. If I continue with letter I get `myVar` caught, I can finish the filling:
![image](https://user-images.githubusercontent.com/1786383/140505832-8edf13ad-1fb8-4ee6-8e48-d7df7c2a645c.png)

3. Barrier which changes snippets to words
    a. `puts $m` as in 1.
    b. `puts $my` similar:
![image](https://user-images.githubusercontent.com/1786383/140506380-c3dba12e-7cce-42b3-8bd5-52ccaeff552f.png)
    c. `puts $myV`:
![image](https://user-images.githubusercontent.com/1786383/140506568-fd174ba6-147b-4813-a7f3-668f37d979a9.png)

For this tests I remove suggestion related settings, so it used defaults.

VS Code version: Code 1.62.0 (b3318bc0524af3d74034b8bb8a64df0ccf35549a, 2021-11-03T15:23:01.379Z)
OS version: Windows_NT x64 10.0.18363
Restricted Mode: No
Remote OS version: Linux x64 3.10.0-1127.10.1.el7.x86_64

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (8 x 2112)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.88GB (8.07GB free)|
|Process Argv|--crash-reporter-id 5677dc62-2d7a-4388-bb68-4cdafcb6b515|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: xxx|
|OS|Linux x64 3.10.0-1127.10.1.el7.x86_64|
|CPUs|Intel(R) Xeon(R) Gold 5218R CPU @ 2.10GHz (80 x 2899)|
|Memory (System)|503.41GB (12.58GB free)|
|VM|0%|
</details><details><summary>Extensions (32)</summary>

Extension|Author (truncated)|Version
---|---|---
bracket-pair-colorizer|Coe|1.0.61
vsc-material-theme|Equ|33.2.0
vsc-material-theme-icons|equ|1.2.2
jupyter-keymap|ms-|1.0.0
remote-ssh|ms-|0.66.0
remote-ssh-edit|ms-|0.66.0
remote-wsl|ms-|0.58.5
material-icon-theme|PKi|4.10.0
jinjahtml|sam|0.16.0
code-settings-sync|Sha|3.4.3
fish-vscode|sky|0.2.1
vscode-icons|vsc|11.7.0
vscode-todo-highlight|way|1.0.5
tcl|bit|0.4.0
prettier-vscode|esb|9.0.0
jenkins-pipeline-linter-connector|jan|1.2.0
rainbow-csv|mec|1.9.1
perforce|mjc|4.15.2
python|ms-|2021.11.1422169775
vscode-pylance|ms-|2021.11.0
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.3
vsliveshare|ms-|1.0.5065
autodocstring|njp|0.5.4
vscode-commons|red|0.0.6
vscode-yaml|red|1.1.0
vscode-sort-json|ric|1.20.0
LiveServer|rit|5.6.1
code-spell-checker|str|2.0.12
vscode-markdown-paste-image|tel|0.13.8
simple-rst|tro|1.5.0
markdown-all-in-one|yzh|3.4.0

(2 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythontb:30283811
pythonvspyt551:30345470
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30384386
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
vsccppwtct:30382698
bridge0708:30335490
pygetstartedt3:30385195
dockerwalkthru:30377721
bridge0723:30353136
pythonrunftest32:30373476
pythonf5test824:30373475
javagetstartedt:30391933
pythonvspyt187:30373474
vsqsis300:30386381
vsaa593:30376534
vssld246:30386377

```

</details>

**Comment by jrieken** [comment]
Can you please share a concrete sample where words should be showing but aren't. What snippet/snippets are taking over in that case? What language is that?

**Comment by maciejmatczak** [comment]
TCL, as mentioned.

Not sure if I can share all snippets, sorry, bit I will try to prepare synthetic test case. Amount of snippets can be important though.

Sorry I didn't do it in the first post, but I hoped something obvious would pop out. That problem just sounds eccentric.

**Comment by maciejmatczak** [comment]
@jrieken, initial issue description updated.

**Comment by jrieken** [comment]
I see this when trigger suggest via `ctrl+space` right after the `$`. Are you saying you don't see that?

<img width="348" alt="Screen Shot 2021-11-05 at 14 19 06" src="https://user-images.githubusercontent.com/1794099/140516398-df01198f-4331-4f1f-8b3f-6d96f7c8911c.png">

**Comment by maciejmatczak** [comment]
That's exactly how it should work, yup.

Moved my `settings.json`, here it what I see:
![image](https://user-images.githubusercontent.com/1786383/140518946-39aadacd-e843-4d72-a679-ce1b081226d7.png)

It's like, whatever was combining the suggestions, got broken.

**Comment by jrieken** [comment]
> It's like, whatever was combining the suggestions, got broken.

Well, there is word based suggestions but no snippet suggestions. There is no snippets because you aren't in-between or inside words but after the `$`. In that case there is no snippets. If you trigger (!) suggest after `$m` you get both - word based suggestions and snippets

**Comment by maciejmatczak** [comment]
You're right, sorry, missed that you typed `m`. But you have an example how it looks at my side in first message. With my `settings.json` moved away nothing changed:
![image](https://user-images.githubusercontent.com/1786383/140520310-54be6658-b272-4ea8-b6cd-99426857ae8f.png)

**Comment by maciejmatczak** [comment]
https://user-images.githubusercontent.com/1786383/140521805-b72b47a9-d2aa-4211-ace8-82fa5e34874b.mp4

- ctrl-space after `$` only triggers word completion, as you described
- adding `m` narrows down choice to one variable
- escaping, retriggering suggestions with `$m` leaves me with snippets only

**Comment by jrieken** [comment]
Can start VS Code like so `code --user-data-dir ~/Desktop/fresh` and try again? With that command you have a truly fresh workspace and zero chance of conflicting settings. Also, make sure you have no settings in your workspace (.vscode-folder)

**Comment by maciejmatczak** [comment]
Tried `--user-data-dir`, didn't help. No `.vscode` folder as well.

`--disable-extensions` makes `TCL` non-functional. Is there a way to just enable one extension?

**Comment by maciejmatczak** [comment]
One more try with `--disable-extensions`, I moved snippets to a global configuration and it seems to work just fine:
![image](https://user-images.githubusercontent.com/1786383/140535028-e3db1293-2664-42d8-8ec7-2cc95dbfbe3e.png)

**Comment by maciejmatczak** [comment]
@jrieken, in a test you made here: https://github.com/microsoft/vscode/issues/136500#issuecomment-961888978

Did you try it more than once? ESC the suggestion, then try again? Delete `m`, then try again re-typing?

It seems that in my production environment, after reload of a window, the first time I type `puts $m`, popup is correct:
![image](https://user-images.githubusercontent.com/1786383/140540995-77bccc14-5998-4278-a8e3-3d147d6190cb.png)

Yet the very next usage drops a word suggestion:
![image](https://user-images.githubusercontent.com/1786383/140541075-2e3dc14a-4059-4ca0-b345-87346f9ac484.png)

**Comment by jrieken** [comment]
> Did you try it more than once? ESC the suggestion, then try again? Delete m, then try again re-typing?

Yes, works repeatedly for me.

**Comment by maciejmatczak** [comment]
I removed Code folder from AppData/Roaming, as well as `.vscode` one from C:/Users/<user>.

I am guessing the latter may have help here, as the first one we covered with additional switch previously. Maybe something else was still pulled from global are though...

`.vscode` one keeps extensions folders, maybe some of them contained some broken data... I still relate this issue with a moment in which VS Code recomputed big amount of snippets for particular language.

**Comment by maciejmatczak** [comment]
After some time bug returned, word based completion still do not work properly for me.

**Comment by maciejmatczak** [comment]
@jrieken, what platform did you use for testing? From your previous help I am guessing it was Linux of some sort.

I am having this issue on Windows, maybe here we have difference.

**Comment by maciejmatczak** [comment]
I tried for an hour to replicate the issue on another windows installation, personal one. No luck, until I tried to install plugins one by one to even it out with my initial setup.

It's `code-spell-checker`. 

Code spell checker not installed:
![obraz](https://user-images.githubusercontent.com/1786383/142066830-2057d4ed-37d0-443a-8f9c-fc1e09edcd89.png)

Code spell checker installed (just in case try triggering suggestions couple of times):
![obraz](https://user-images.githubusercontent.com/1786383/142066955-4f1d310d-76cd-44d3-a84c-37dfe184f4a2.png)
